### ISSUE TITLE
Scene Setup

### ISSUE THREAD

**Issue by tokejepsen** [getavalon/core#331]
**Goal**

Prompt users for fixing the scene setup before doing any work.

**Motivation**

A lot of setup is required before the artist starts work to prevent loss of work. Examples are;

- FPS
- Frame Range
- Resolution
- Work Directory
- Others?

This has been implemented in various configs, but since its a common problem I would suggest to tackle this in the core.

**Implementation**

Currently hosts (Maya) has a common method for setting the frame range and resolution to what is configured in the database. This however needs to be explicitly run by the users, which can lead to forgetting the setup steps.
Instead of explicitly opting in, I would suggest explicitly opting out. By default the scene would be setup to match the database, unless the user cancels the operation.

Scene setup needs to happen in three areas;

- when the user initializes a new scene.
- when the user opens a scene.
- when the user switches context.

An initial suggestion for a GUI would be a simple message box that states what is out of sync with the database, and what happens when pressing "ok". If the user presses "ok" the scene setup will be synced with the database.
This GUI would need to be very visible to the user and maybe even modal, so the user has to make a decision before starting any work.

Unfortunately not all applications will work the same, so we can't make cross application assumptions. For example NukeStudio will not to setup FPS/Frame Range/Resolution but might need Work Directory.

**Comment by BigRoy** [comment]
We had a similar discussion internally. My final statement about that was:

I'd never want the scene to be altered without my consent as an Artist. I would not want to magically have my FPS altered without my knowledge, similarly I don't want it to be automatically "fixed" upon a render. ;)

What I do like is to be notified in such a way that it doesn't hold me back if I want to disregard it yet do want to clearly indicate the visual warning that the user won't miss.

The tricky thing with a warning message is that scene open can take some time and the user will get the message afterwards. If a scene takes e.g. >10 seconds to load Artists often do other things in between. Meaning that, when coming back afterwards, the warning should still be clearly visible.

In short, in a way it's more about getting the right visual cue. Whether it states "fix your fps?" with "ok" and "ignore" or it would state "would you like to match your fps to shot data?" with "yes" ans "no" I wouldn't care too much about.

It is however important that when fixing it it is clear to the artist what the current state is and what it will become? E.g. 24fps -> 25fps.

Also I felt that modal dialogs (holding maya from continuing) were somewhat frustrating since it held me back.

Does that input help?

**Comment by tokejepsen** [comment]
> Does that input help?

Think we are on the same page :)

**Comment by tokejepsen** [comment]
https://medium.com/blue-sky-tech-blog/introducing-achievements-into-qc-6fee943324df

Blue Sky has a nice term for this called "Preflight" validation. Might be a nice name for a tool 😀

**Comment by tokejepsen** [comment]
We could utilize the validation part of Pyblish core library for this.
Just need to have plugins targeted to an agreed target name like "preflight" or similar.

**Comment by tokejepsen** [comment]
@BigRoy have you got anywhere with this?

I might be looking to make something for this pretty soon.

Also @mkolar has some insights about this on the Pype pipeline?

**Comment by mkolar** [comment]
We only made a rudimentary version for nuke so far that gathers latest relevant publishes, loads them all and connects what is possible. Not very robust, but it is a start. I'll push latest pype to github at the end, where it can be visible.

**Comment by mkolar** [comment]
>What Milan describes does sound different from what the issue itself describes originally, but I can see how that might fall under scene setup. However what he describes I think is better fitted with #130 but I might be mistaken.

Aah you're completely right, I haven't read properly.
